contracts-js/BTCUtils.js

```diff
@@ -0,0 +1,1012 @@
+//     /// @dev             Use verifyHash256Merkle to verify proofs with this root
+//     /// @param _header   The header
+//     /// @return          The merkle root (little-endian)
+//     function extractMerkleRootLE(bytes memory _header) internal pure returns (bytes memory) {
+//         return _header.slice(36, 32);
+//     }
+
+  /// @notice          Extracts the transaction merkle root from a block header
+  /// @dev             Use verifyHash256Merkle to verify proofs with this root
+  /// @param _header   The header
+  /// @return          The merkle root (little-endian)
+  extractMerkleRootLE: (header) => {
+    return
+  },
+
+//     /// @notice          Extracts the transaction merkle root from a block header
+//     /// @dev             Use verifyHash256Merkle to verify proofs with this root
+//     /// @param _header   The header
+//     /// @return          The merkle root (big-endian)
+//     function extractMerkleRootBE(bytes memory _header) internal pure returns (bytes memory) {
+//         return reverseEndianness(extractMerkleRootLE(_header));
+//     }
+
+  /// @notice          Extracts the transaction merkle root from a block header
+  /// @dev             Use verifyHash256Merkle to verify proofs with this root
+  /// @param _header   The header
+  /// @return          The merkle root (big-endian)
+  extractMerkleRootBE: (header) => {
+    return
+  },
+
+//     /// @notice          Extracts the target from a block header
+//     /// @dev             Target is a 256 bit number encoded as a 3-byte mantissa and 1 byte exponent
+//     /// @param _header   The header
+//     /// @return          The target threshold
+//     function extractTarget(bytes memory _header) internal pure returns (uint256) {
+//         bytes memory _m = _header.slice(72, 3);
+//         uint8 _e = uint8(_header[75]);
+//         uint256 _mantissa = bytesToUint(reverseEndianness(_m));
+//         uint _exponent = _e - 3;
+
+//         return _mantissa * (256 ** _exponent);
+//     }
+
+  /// @notice          Extracts the target from a block header
+  /// @dev             Target is a 256 bit number encoded as a 3-byte mantissa and 1 byte exponent
+  /// @param _header   The header
+  /// @return          The target threshold
+  extractTarget: (header) => {
+    let d_header = utils.deserializeHex(header)
+
+    // Hacky way of reversing endianness of a partial serialized number
+    let m = header.slice(72, 75).split().reverse().join('')
+    let hex_m = `0x${m}`
+
+    let e = d_header[75] - 3
+    let exponent = BigInt(256 ** (e - 3)) // FIX: throws an unsafe number error
+    let mantissa = utils.bytesToUint(hex_m)
+
+    console.log('mantissa: ', mantissa)
+    // console.log('e: ', e)
+    // console.log('m: ', m)
+    // let exponent = BigInt(256 ** (e-3))
+    // console.log('exponent: ', exponent)
+    // let exponent = 256 ** (e - 3)
+
+    return mantissa * exponent
```

Comment:
```
let exponent = (e - 3n)
return mantissa * 256n ** exponent
```

Reply:
Made changes to previous comments.
I guess node just doesn't understand BigInt. It throws a TypeError: `Do not know how to serialize a BigInt`. This occurs when something is trying to use JSON.stringify on the result.

Reply:
> This occurs when something is trying to use JSON.stringify on the result.

Ya it's not ideal. `JSON.stringify` will probably be broken for them for a while

Reply:
k